### PR TITLE
Improvements 1a

### PR DESCRIPTION
**Pull request by edbeeching**
This PR:
- [x] Cleans up setup.py and remove settings.ini (legacy from nbdev)
- [x] Makes all internal imports relative (e.g. from trl.trainer -> from .trainer)
- [x] Cleans up README and fixes the example

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -61,57 +49,70 @@ This is a basic example on how to use the library. Based on a query the language
...
+from transformers import AutoTokenizer
+from trl import PPOTrainer, PPOConfig, AutoModelForCausalLMWithValueHead
+from trl.core import respond_to_batch
 
 # get models
-gpt2_model = GPT2HeadWithValueModel.from_pretrained('gpt2')
-gpt2_model_ref = GPT2HeadWithValueModel.from_pretrained('gpt2')
-gpt2_tokenizer = GPT2Tokenizer.from_pretrained('gpt2')
+gpt2_model = AutoModelForCausalLMWithValueHead.from_pretrained('gpt2')
+gpt2_model_ref = AutoModelForCausalLMWithValueHead.from_pretrained('gpt2')
```
Comment (lvwerra) [inline README.md:58]: we can now use:

```suggestion
gpt2_model_ref = create_refrence_model(gpt2_model)
```

or we can not provide a reference model at all in which case the trainer will call above functions.

File: README.md [hunk 2]
```diff
@@ -61,57 +49,70 @@ This is a basic example on how to use the library. Based on a query the language
...
+    forward_batch_size=1
+)
 
 # encode a query
 query_txt = "This morning I went to the "
 query_tensor = gpt2_tokenizer.encode(query_txt, return_tensors="pt")
 
 # get model response
-response_tensor  = respond_to_batch(gpt2_model, query_tensor)
+response_tensor  = respond_to_batch(gpt2_model_ref, query_tensor)
```
Comment (lvwerra) [inline README.md:72]: Why did you change this? I think the response should be given by the active model, no?

File: README.md [hunk 3]
```diff
@@ -61,57 +49,70 @@ This is a basic example on how to use the library. Based on a query the language
...
+        self.query_data = query_data
+        self.response_data = response_data
+
+    def __len__(self):
+        return len(self.query_data)
+
+    def __getitem__(self, idx):
+        return self.query_data[idx], self.response_data[idx]
+
+dummy_dataset = torch.utils.data.Dataset()
```
Comment (lvwerra) [inline README.md:87]: This seems unnecessary?

File: README.md [hunk 4]
```diff
@@ -61,57 +49,70 @@ This is a basic example on how to use the library. Based on a query the language
...
-from trl.ppo import PPOTrainer
+from transformers import AutoTokenizer
+from trl import PPOTrainer, PPOConfig, AutoModelForCausalLMWithValueHead
+from trl.core import respond_to_batch
 
 # get models
-gpt2_model = GPT2HeadWithValueModel.from_pretrained('gpt2')
-gpt2_model_ref = GPT2HeadWithValueModel.from_pretrained('gpt2')
-gpt2_tokenizer = GPT2Tokenizer.from_pretrained('gpt2')
+gpt2_model = AutoModelForCausalLMWithValueHead.from_pretrained('gpt2')
```
Comment (lvwerra) [inline README.md:57]: What do you think about omitting `gpt2_` from tokenizer and model names?

File: README.md [hunk 5]
```diff
@@ -61,57 +49,70 @@ This is a basic example on how to use the library. Based on a query the language
...
+)
+
+# create a ppo trainer
+ppo_trainer = PPOTrainer(ppo_config, gpt2_model, gpt2_model_ref, gpt2_tokenizer, dummy_dataset)
+device = ppo_trainer.accelerator.device
+
 # define a reward for response
 # (this could be any reward such as human feedback or output from another model)
-reward = [torch.tensor(1.0)]
+reward = [torch.tensor(1.0).to(device)]
```
Comment (lvwerra) [inline README.md:102]: We should let `accelerate` handle device placement internally. (Can address it in a follow up PR if you want)

File: setup.py [hunk 6]
```diff
@@ -1,51 +1,47 @@
...
+    license="Apache 2.0",
+    classifiers=[
+            "Development Status :: 2 - Pre-Alpha",
+            "Intended Audience :: Developers",
+            "Intended Audience :: Science/Research",
+            "License :: OSI Approved :: Apache Software License",
+            'Natural Language :: English',
+            "Operating System :: OS Independent",
+            "Programming Language :: Python :: 3",
+            "Programming Language :: Python :: 3.6",
```
Comment (lvwerra) [inline setup.py:27]: Since we don't test it anymore we could also remove it here, right?
```suggestion
```

File: trl/trainer/__init__.py [hunk 7]
```diff
@@ -14,7 +14,11 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+# isort: off
+from .utils import AdaptiveKLController, FixedKLController
+
+# isort: on
+
```
Comment (lvwerra) [inline trl/trainer/__init__.py:21]: Why are the utils imports separate from the others?

**Review by lvwerra — COMMENTED** [review]
Thanks @edbeeching, this looks great! Left a few minor comments, can also address some of them in a follow up PR.

File: README.md (hunk 2 above)
Comment (edbeeching) [reply README.md:72]: Yes you are right, in this case the model is just used to create the data for the dummy dataset, but I will update it because in a normal training setting it would not be like this.

File: README.md (hunk 5 above)
Comment (edbeeching) [reply README.md:102]: I think the reward would normally be from a preference model, so it would probably be on the correct device. But in this example that is not the case. I'll leave it to another PR.

Comment (lvwerra) [reply README.md:100]: yes, discussed with @younesbelkada that we add some device placement steps to `PPOTrainer.step`.

**Review by lvwerra — APPROVED** [review]
Looks good - thanks @edbeeching!

**Review by younesbelkada — APPROVED** [review]
Thanks for taking care of the improvements @edbeeching !

File: README.md (hunk 5 above)
Comment (younesbelkada) [reply README.md:100]: yes, this will be addressed in a PR right after merging this
